metrics: Metrics initial modularization

dylandreimerink opened the pull request:
This PR is the first step towards full modularization of metrics. It is a smaller version of #23936 to make review easier. This PR makes the registry into a cell, yet still registers this cell as global variable so the global functions stay working. So in essence we are in a sort of hybrid mode so we do not have to modify as much metric call sites.

This version adds the `cell.Metric` cell-type which allows developers to define metrics wherever they want. This type works just like a `cell.Provide`, you give it a constructor which returns a struct containing metrics. The main difference is that both the struct and its fields are provided to the DAG so the registry can collect the individual metrics while normal cells can use the struct as param and have named metrics.

To provide support for disabled-by-default metrics, we added wrapper types around existing prometheus types, defined in `pkg/metrics/metric`. They are a super-set of their normal counterparts so they don't break the existing call sites. Having this config inside the metrics options at creation time eliminates the need for the "is enabled" feedback to the central config. So the metrics are not fully standalone without dependency on the Daemon it init or the Daemon config.

All existing metrics are converted to these new metrics and are made available to Hive via a new `LegacyMetrics` struct. They are also still registered at the old global variables so the call sites do not have to be updated in this PR. Follow-up PRs should incrementally remove usage of these global variables in favor of taking LegacyMetrics as parameter or moving metrics of interest into a dedicated metrics struct closer to the call site.

Follow-up PRs should also enrich the metrics type to allow for self documenting functionality like the original PR and all dynamic registration/unregistration should be refactored to use the registry as parameter or to make the collectors smarter so we can move away from global functions to modify the registry.

dylandreimerink committed:
hive,metrics: Add `cell.Metric` and `metrics/metric` package

This commit adds the metric cell type which can be used to make metrics
available to hive. The cell works by giving it a constructor function
just like providing a normal type. The returned type should contain
public metric fields, the cell will enforce this. The cell will make
both the populated metric struct as well as all of the individual
metrics available in the DAG. The idea being that users would use the
struct so they can access just the metrics they are interested in. The
metrics registry will collect all individual metrics and register them.

We currently have the concept of metrics being enabled/disabled by
default unless configured otherwise. To support this concept in a
modular way we need additional metadata for metrics. Therefor all
metrics should be defined with the new `metric` package which contains
types that wrap existing prometheus metrics and it adds our own options
such as `Disabled` and `ConfigName`.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

dylandreimerink committed:
metrics: Modularize daemon metrics registry

This commit modularizes the metrics package. We now have a registry
cell which collects all metrics exposed via `cell.Metric` and registers
them with an internal prometheus registry. This registry operates
independently now, the HTTP server is part of its lifecycle and the
daemon and daemon config are no longer involved directly.

We still have some metric types which implement dynamic registration
and logic external to the metrics package. The global functions now
await a promise to the Registry so they still work until we can replace/
refactor these external dynamic metrics.

All of the metrics defined in the metrics package have been transformed
into `metrics/metric` types and are make available via a Metrics cell
called `LegacyMetics`. This cell will also assign the individual metrics
to the existing global variables so they can be used until the callsites
can be adjusted to take `LegacyMetrics` as dependency or to split off
some of the metrics contained inside.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

bimmlerd commented on pkg/hive/example/metrics.go:
```diff
@@ -0,0 +1,28 @@
+	"github.com/cilium/cilium/pkg/metrics/metric"
+)
+
+var exampleMetricsCell = cell.Metric(newExampleMetrics)
```
I believe the the expected use case of metrics cells to be part of a module; maybe the example could be extended slightly to show that pattern?

dylandreimerink replied:
> I believe the the expected use case of metrics cells to be part of a module

That is not necessary... what gave you the impression?

bimmlerd replied:
I guess I expect metrics to be associated with cells, typically. Not as a requirement, just as the usual way of doing things - if that's not the case, feel free to resolve this.

bimmlerd commented on pkg/hive/example/metrics.go:
```diff
@@ -0,0 +1,28 @@
+// SPDX-License-Identifier: Apache-2.0
```
[commenting here for lack of a better place]
Could we add a section to the "Guide to the hive" docs which explains these modular metrics in this first commit?

dylandreimerink replied:
Ah, I always forget the docs. Will do!

bimmlerd commented on pkg/metrics/metric/metric.go:
```diff
@@ -0,0 +1,93 @@
+	Reset()
+}
+
+// Opts are a modified and extended version of the prometheus.Opts
```
I can only see extensions (`ConfigName`, `EnabledByDefault` are added). If that's the case; would it not be better to embed a prometheus.Opts?

dylandreimerink replied:
There are two reasons for not doing so. First is that I expect to change the types of `Subsystem` in a later PR which will make embedding impossible. Secondly, making literals with embedded objects is a pain:

```
opt := NoEmbeds{
  Name: "",
  Namespace: "",
  EnabledByDefault: true,
}

opt2 := WithEmbeds{
  opts: prometheus.Opts{
    Name: "",
    Namespace: "",
  },
  EnableByDefault: true,
}
```

bimmlerd replied:
Ack re the first reason. The reason why I asked is because we'd get future changes in prometheus.Opts for free if we embed; as is the maintenance is potentially increased (?).

bimmlerd commented on pkg/metrics/metric/histogram.go:
```diff
@@ -0,0 +1,331 @@
+}
+
+// HistogramOpts are a modified and expanded version of the prometheus.HistogramOpts.
+// https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#HistogramOpts
```
Similar question here, why not embed - I only see extensions.

dylandreimerink replied:
See answer above

bimmlerd commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -242,7 +242,7 @@ cilium-agent [flags]
       --mesh-auth-spiffe-trust-domain string                    The trust domain for the SPIFFE identity. (default "spiffe.cilium")
       --mesh-auth-spire-admin-socket string                     The path for the SPIRE admin agent Unix socket.
-      --metrics strings                                         Metrics that should be enabled or disabled from the default metric list. The list is expected to be separated by a space. (+metric_foo to enable metric_foo , -metric_bar to disable metric_bar)
+      --metrics strings                                         Metrics that should be enabled or disabled from the default metric list. (+metric_foo to enable metric_foo , -metric_bar to disable metric_bar)
```
nit: is the space after `metric_foo` intentional? (unchanged from existing, just curious)

dylandreimerink replied:
These files are auto generated, no idea why its different.

bimmlerd commented on daemon/cmd/daemon_test.go:
```diff
@@ -43,6 +40,8 @@ import (
 	"github.com/cilium/cilium/pkg/testutils"
 	"github.com/cilium/cilium/pkg/types"
+
+	. "github.com/cilium/checkmate"
```
nit: is this a goimports thing?

dylandreimerink replied:
Yes it is, for some reason goimports wants it below once the prometheus is gone

bimmlerd commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -242,7 +242,7 @@ cilium-agent [flags]
       --mesh-auth-rotated-identities-queue-size int             The size of the queue for signaling rotated identities. (default 1024)
```
2nd commit msg nit: "independency" -> "independently"

bimmlerd commented on daemon/cmd/endpoint_test.go:
```diff
@@ -146,6 +147,7 @@ func assertOnMetric(c *C, state string, expected int64) {
 		return obtained == expected
 	}, 10*time.Second)
 	if err != nil {
+		spew.Dump(metrics.EndpointStateCount, state)
```
nit: leftover from debugging or intentional?

bimmlerd commented on pkg/metrics/registry.go:
```diff
@@ -0,0 +1,242 @@
+type RegistryConfig struct {
+	// PrometheusServeAddr IP:Port on which to serve prometheus metrics (pass ":Port" to bind on all interfaces, "" is off)
+	PrometheusServeAddr string
+	//
```
nit: empty comment

bimmlerd commented on pkg/metrics/registry.go:
```diff
@@ -0,0 +1,242 @@
+			continue
+		}
+
+		switch metricFlag[0] {
```
I think we should add a default case with a warning to this switch, to prevent silently losing some metrics - I _think_ the following would lose the middle metric disable (due to the space before it), as well as silently ignoring the final metric (because there's no plus/minus):

`--metrics "+first_metric, -middle_metric,final_metric"`

I traced `viper.StringSlice` to `encoding/csv` [which doesn't trim whitespace](https://go.dev/play/p/fok-8A8Sdg9).
Given this has been the case in the existing code, probably error/fatal would be breaking and can't be done.

bimmlerd commented on pkg/metrics/registry.go:
```diff
@@ -0,0 +1,242 @@
+	}
+
+	for _, val := range currentMetrics {
+
```
nit: spurious newline

bimmlerd commented on pkg/endpoint/endpoint_test.go:
```diff
@@ -77,15 +73,12 @@ func (s *EndpointSuite) SetUpSuite(c *C) {
-	for _, c := range s.collectors {
-		metrics.Unregister(c)
-	}
+	metrics.EndpointStateCount = metrics.NoOpGaugeVec
```
How does this differ from `SetEnabled(false)`? Just wondering about the asymmetry with above

dylandreimerink replied:
Originally, the code would create just the `EndpointStateCount` metric, and register it at the prometheus registry, initializing its global variable in the process. Then they unregister the metric from the registry, in case a second tests want to register the same metric, which would not be allowed.

But in this whole exchange, the registry is never actually used, just the metric. So what I do instead is only create the metric, don't bother with the registry, and at the end reset the global variable back to the value before starting this test.

dylandreimerink replied:
I think `SetEnabled(true)` would have also worked here btw, I can change that.

bimmlerd replied:
Either way works, my OCD just likes symmetry :sweat_smile:

bimmlerd commented on pkg/endpointmanager/manager.go:
```diff
@@ -192,10 +192,11 @@ func (mgr *endpointManager) InitMetrics() {
+			Namespace:        metrics.Namespace,
+			Name:             "endpoint",
+			Help:             "Number of endpoints managed by this agent",
+			EnabledByDefault: true,
```
Should we, by Go principles of making the empty value useful, instead be using `Disabled`, defaulting to `true`?

dylandreimerink replied:
That is a good question. I made it like this because the original pattern was disabled by default unless specifically enabled. I can go either way.

bimmlerd replied:
[I mean to write "defaulting to `false`" in the first comment]

I don't have a strong opinion. If it's not a large change, I think it's a little nicer.

dylandreimerink replied:
Agreed, but if the field is "Disabled" and you don't specify it, then the metric is enabled by default. Which is the opposite of what it was before. I personally like your suggestion better since we have more metrics enabled by default anyways. But perhaps someone from @cilium/metrics should way in

chancez replied:
In practice, most metrics should be enabled by default, so I would lean towards `Disabled`, as that would promote metrics being enabled by default. It's okay if some metrics are just `Disabled: true` by default, especially if they're for specific use-cases or have high cardinality.

bimmlerd commented on pkg/hive/example/main.go:
```diff
@@ -15,10 +15,11 @@ import (
+		eventsCell,         // Example event source (ExampleEvents)
+		helloHandlerCell,   // Handler for /hello
+		eventsHandlerCell,  // Handler for /events
+		exampleMetricsCell, // Register the metrics
```
I think this topically belongs to the first commit

dylandreimerink replied:
It was supposed to, must have made a mistake somewhere.

bimmlerd commented on pkg/metrics/registry.go:
```diff
@@ -0,0 +1,242 @@
+
+func (rc RegistryConfig) Flags(flags *pflag.FlagSet) {
+	flags.String("prometheus-serve-addr", rc.PrometheusServeAddr, "IP:Port on which to serve prometheus metrics (pass \":Port\" to bind on all interfaces, \"\" is off)")
+	flags.StringSlice("metrics", rc.Metrics, "Metrics that should be enabled or disabled from the default metric list. (+metric_foo to enable metric_foo , -metric_bar to disable metric_bar)")
```
github doesn't let me do the suggestion thing, but the spurious space comes from here and can go :)

bimmlerd reviewed (commented):
Thanks a bunch for making this easier to review! Have a couple of comments, but nothing major, awesome work!

bimmlerd reviewed (approved):
LGTM with the understanding that docs will be added. Cool stuff!

nathanjsweet commented on clustermesh-apiserver/metrics/metrics.go:
```diff
@@ -36,7 +35,7 @@ type MetricsConfig struct {
 
 func (def MetricsConfig) Flags(flags *pflag.FlagSet) {
-	flags.String(option.PrometheusServeAddr, def.PrometheusServeAddr, "Address to serve Prometheus metrics")
+	flags.String("prometheus-serve-addr", def.PrometheusServeAddr, "Address to serve Prometheus metrics")
```
Can we keep the option string a constant?

tommyp1ckles commented on pkg/bpf/bpf_linux.go:
```diff
@@ -58,7 +57,7 @@ type bpfAttrMapOpElem struct {
 func UpdateElementFromPointers(fd int, mapName string, structPtr unsafe.Pointer, sizeOfStruct uintptr) error {
 	var duration *spanstat.SpanStat
-	if option.Config.MetricsConfig.BPFSyscallDurationEnabled {
+	if metrics.BPFSyscallDuration.IsEnabled() {
```
q: Do we even need to check here, if its a no-op if its not enabled?

dylandreimerink replied:
We technically don't need the check, if not enabled functions like `Inc`, `Add`, `Observe` are no-op. This was the case before as well so I assume this is an optimization to avoid the cycles used to collect the measurement if we are not going to use it anyway.  No clue if that is warranted or not.

tommyp1ckles commented on pkg/hive/cell/metric.go:
```diff
@@ -0,0 +1,139 @@
+		))
+	}
+
+	var withMeta pkgmetric.WithMetadata
```
nit: could we do this as a package scoped variable once instead?

tommyp1ckles commented on pkg/hive/cell/metric.go:
```diff
@@ -0,0 +1,139 @@
+	collectorTyp := reflect.TypeOf(&collector).Elem()
+	for i := 0; i < outTyp.NumField(); i++ {
+		field := outTyp.Field(i)
+		if !field.IsExported() {
```
Why do we need this constraint on metrics, I don't think other cells enforce this?

dylandreimerink replied:
If a field containing a metric isn't exported, we can't make it separately available in Hive. That would result in the odd behavior where the package defining the metric struct can still access, modify and use the metric but it will never be exported. That would happen silently, so by enforcing this I hope to avoid users making this hard to find bug and easy to overlook type of bug.

tommyp1ckles commented on pkg/hive/cell/metric.go:
```diff
@@ -0,0 +1,139 @@
+	pkgmetric "github.com/cilium/cilium/pkg/metrics/metric"
+)
+
+func Metric[S any](ctor func() S) Cell {
```
Given all the runtime constraints on the passed constructor, might be worth documenting this in the function comment.

dylandreimerink committed:
docs: Added docs for new metric cell

This commits adds a new section to the "Guide to Hive" page explaining
the new `cell.Metric` cell type.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

giorio94 commented on pkg/hive/cell/metric.go:
```diff
@@ -0,0 +1,152 @@
+
+	"go.uber.org/dig"
+
+	"github.com/prometheus/client_golang/prometheus"
```
Very minor nit. These two imports should be likely merged in a single block.

dylandreimerink replied:
I really need to look into the goimports rules, I am not making this shit up myself, just the result of running ` goimports -w -local github.com/cilium/cilium <file>` :man_shrugging:

giorio94 replied:
IIRC it does preserve empty lines, which is kinda weird. [gci](https://github.com/daixiang0/gci) should be stricter from that point of view.

dylandreimerink replied:
Anyway, its fixed now

giorio94 commented on pkg/hive/cell/metric.go:
```diff
@@ -0,0 +1,152 @@
+	if outTyp.Kind() != reflect.Struct {
+		panic(fmt.Sprintf(
+			"cell.Metric must be invoked with a constructor function that returns a struct or pointer to a struct, "+
+				"%s supplied a constructor which returns a %d",
```
Nit. I guess `%d` should be `%s`.

giorio94 commented on pkg/metrics/metric/metric.go:
```diff
@@ -0,0 +1,93 @@
+	ConfigName string
+
+	// If true, the metric has to be explicitly enabled via config or flags
+	Disabled bool
```
Nit. The commit message still mentions `EnabledByDefault` rather than `Disabled`.

giorio94 commented on pkg/metrics/metric/metric.go:
```diff
@@ -0,0 +1,93 @@
+	GetMetricWith(labels prometheus.Labels) (T, error)
+	GetMetricWithLabelValues(lvs ...string) (T, error)
+	With(labels prometheus.Labels) T
+	WithLabelValues(lvs ...string) T
```
I wonder whether it could make sense to propagate the comment of these methods, for reference. Same for the ones below.

dylandreimerink replied:
I added the comments to the interface and modified it to fit with the generic interfaces. I think this will end up neater in code completion and generated docs.

giorio94 commented on pkg/metrics/metric/counter.go:
```diff
@@ -0,0 +1,179 @@
+	prometheus.Counter
+	WithMetadata
+
+	Get() float64
```
Out of curiosity, is `Get` provided for test purposes only, or is it there another reason? In that case, is it there any reason to prefer it over the `"prometheus/testutil"` package?

dylandreimerink replied:
Its meant to replace `metrics.GetCounterValue` and `metrics.GetGaugeValue` in future PRs. These are used both in tests and the metricsmap which is a bit of an oddball. 

I never knew `prometheus/testutil` existed, just thought I would replace the above with a nice methods since we are wrapping anyway. I have no personal preference.

giorio94 replied:
I don't have a strong preference either. My only concern is that if being part of the interface they might encourage a bit more incorrect usages outside of testing.

giorio94 commented on pkg/metrics/metric/counter.go:
```diff
@@ -0,0 +1,179 @@
+
+func (c *counter) Opts() Opts {
+	return c.opts
+}
```
Nit. It might be moved to the `metric` common implementation.

giorio94 commented on pkg/metrics/metric/gauge.go:
```diff
@@ -0,0 +1,237 @@
+	gv.metric.SetEnabled(e)
+}
+
+type GaugeFunc interface {
+	prometheus.GaugeFunc
+	WithMetadata
+}
```
`Collect()` doesn't seem to be aliased for this metric type.

giorio94 commented on pkg/metrics/metric/gauge.go:
```diff
@@ -0,0 +1,237 @@
+
+func (g *gauge) Describe(descChan chan<- *prometheus.Desc) {
+	g.Gauge.Describe(descChan)
+}
```
Nit. `Describe()` is not strictly required to be aliased given that `prometheus.Gauge` is embedded (same for the other metric types).

giorio94 commented on pkg/hive/example/hello_handler.go:
```diff
@@ -16,11 +16,12 @@ var helloHandlerCell = cell.Module(
 )
 
-func newHelloHandler() HTTPHandlerOut {
+func newHelloHandler(ExampleMetrics exampleMetrics) HTTPHandlerOut {
```
I guess these changes should be part of the previous commit.

giorio94 commented on pkg/metrics/middleware_test.go:
```diff
@@ -35,8 +36,10 @@ func (s *MetricsSuite) TestAPIEventsTSHelperMiddleware(c *C) {
-		gauge := prometheus.NewGaugeVec(prometheus.GaugeOpts{}, []string{LabelEventSource, LabelScope, LabelAction})
-		hist := prometheus.NewHistogramVec(prometheus.HistogramOpts{Name: "test_api_hist"}, []string{LabelEventSource, LabelScope, LabelAction})
+		gauge := metric.NewGaugeVec(metric.GaugeOpts{}, []string{LabelEventSource, LabelScope, LabelAction})
+		gauge.SetEnabled(true)
```
Nit `gauge.SetEnabled` is superfluous as enabled by default.

giorio94 commented on pkg/metrics/metric/metric.go:
```diff
@@ -8,10 +8,8 @@ import (
-// documentation from de definitions. In addition, these metrics have the concept of being enabled or disabled which
-// is used in place of conditional registration so all metric types can always be registered for the purposes of
-// documenting them.
+// prometheus metric types and add additional metadata. In addition, these metrics have the concept of being enabled
+// or disabled which is used in place of conditional registration so all metric types can always be registered.
```
Nit. I guess that this change should belong to the previous commit.

giorio94 commented on pkg/metrics/registry.go:
```diff
@@ -0,0 +1,249 @@
+
+	for _, m := range metrics {
+		if c, ok := m.(prometheus.Collector); ok {
+			r.MustRegister(c)
```
Previously, the `Describe()` function of NoOp metrics was a no-op, while AFAICT it now propagates the description regardless of whether the metric is enabled or not. I assume that this makes prometheus report the metric, even though with zero value. Is this change intentional?

dylandreimerink replied:
Did a dive into the code to check how this works and a test to confirm. The `Describe()` is only used to check the options of the metric at registration time (and perhaps does some additional runtime checks for consistency). But if the `Collect()` does not return any metrics, no entries will appear on `/metrics` not even zero values. So this does what we expect it to do. Plus it adds some additional collision detection even for Disabled by default metrics.

giorio94 commented on clustermesh-apiserver/metrics/metrics.go:
```diff
@@ -36,7 +35,7 @@ type MetricsConfig struct {
 
 func (def MetricsConfig) Flags(flags *pflag.FlagSet) {
-	flags.String(option.PrometheusServeAddr, def.PrometheusServeAddr, "Address to serve Prometheus metrics")
+	flags.String("prometheus-serve-addr", def.PrometheusServeAddr, "Address to serve Prometheus metrics")
```
clustermesh-apiserver metrics are no longer exported, because they are no longer initialized: https://github.com/cilium/cilium/blob/a7a483ac3782574e56d63e7adbd5d6fb61686743/clustermesh-apiserver/metrics/metrics.go#L67-L73 I think it should be enough to make the metrics cell depend on the `LegacyMetrics` struct.

dylandreimerink replied:
Making `LegacyMetrics` a dependency requires that `metrics.Cell` is included in the clustermesh-apiserver hive, but that would cause a collision since they both register this same flag.

For now I solved it by just calling the constructor without hive. In the future, once these KVStore metrics have their own metric cell we can make the clustermesh-apiserver "just" re-use the `metrics.Registry` without the legacy metrics but include the KVStore metric cell in the hive. That way we have 1 copy of the registry logic but different hives include different metrics. The same goes for the operator metrics

giorio94 reviewed (commented):
Overall this PR looks awesome to me! I've left only a bunch of mostly minor comments and nits inline.

lmb commented on daemon/cmd/daemon_test.go:
```diff
@@ -138,18 +134,9 @@ func (epSync *dummyEpSyncher) DeleteK8sCiliumEndpointSync(e *endpoint.Endpoint)
-		metrics.Unregister(c)
-	}
-}
+func (ds *DaemonSuite) TearDownSuite(c *C) {}
```
Nit: why not just get rid of this?

dylandreimerink replied:
Honestly, I am not that familiar with the whole suite tests, I had assumed you can't just remove functions without consequences, will try that :smile:

lmb commented on pkg/bpf/map_linux.go:
```diff
@@ -279,7 +279,7 @@ func (m *Map) updatePressureMetric() {
 	// Do a lazy check of MetricsConfig as it is not available at map static
 	// initialization.
-	if !option.Config.MetricsConfig.BPFMapPressure {
+	if !metrics.BPFMapPressure {
```
Why is this not `.IsEnabled()`?

dylandreimerink replied:
Well because the `.IsEnabled()` is a feature of the new `pkg/metrics/metric` types, and the map pressure metrics are implemented with their own custom type `metrics.GaugeWithThreshold`. To keep this PR manageable, I decided to not refactor it, will do that in a followup PR.

lmb commented on pkg/hive/cell/metric.go:
```diff
@@ -0,0 +1,151 @@
+func Metric[S any](ctor func() S) Cell {
+	caller := "caller"
+	if _, file, no, ok := runtime.Caller(1); ok {
+		caller = fmt.Sprintf("%s:%d", file, no)
```
Why manually annotate with the caller if the function panics anyways?

dylandreimerink replied:
Yea, that makes sense

lmb commented on pkg/hive/cell/metric.go:
```diff
@@ -0,0 +1,151 @@
+		outTyp = outTyp.Elem()
+	}
+
+	if outTyp.Kind() != reflect.Struct {
```
Meta comment that isn't specific to this PR, more of a discussion. Any time we design an API where the constraints can only be enforced at runtime we lose developer productivity / velocity.

dylandreimerink replied:
Very good point. I try to use types wherever I can to make it clear to users what can and can't be used at compile time. But at this moment we just do not have that capability to the depth I would like.

In this case, we could also have gone for a `cell.Config` like model where the generic type must implement a given interface and have the interface return a slice of metrics. The very first iteration worked like this, but it requires the user to write a lot of extra boilerplate which I disliked.

At the moment I feel like I have to pick between two evils, this being the lesser from my point of view. But I am open to discussing this further, perhaps on slack somewhere.

lmb commented on pkg/hive/cell/metric.go:
```diff
@@ -0,0 +1,151 @@
+
+	value := reflect.ValueOf(metricSet)
+	typ := value.Type()
+	if typ.Kind() == reflect.Pointer {
```
You can use https://go.dev/pkg/reflect/?m=old#Indirect to avoid this

dylandreimerink replied:
I tries, but that only works on a `reflect.Value` not on a `reflect.Type`.

lmb commented on pkg/hive/cell/metric.go:
```diff
@@ -0,0 +1,151 @@
+// value available via the `hive-metrics` value group. Infrastructure components such as a registry,
+// inspection tool, or documentation generator can collect all metrics in the hive via this value group.
+//
+// The `ctor` constructor must return a struct or pointer to a struct of type `S`. The returned struct
```
Does it make sense to return non-pointer structs at all? I'm guessing that metrics structs will be on the larger side.

dylandreimerink replied:
It depends. The fields of these structs will all be pointers, so its not that big of a deal. Besides, they are mostly transferred during construction, after that they will likely be part of the field of a cell somewhere. I don't think we should restrict it any further if we can help it.

bimmlerd commented on daemon/cmd/cells.go:
```diff
@@ -60,6 +61,8 @@ var (
 
 		cni.Cell,
 
```
nit: comment to explain the purpose of this cell?

bimmlerd reviewed (approved):
contributing part lgtm too

qmonnet commented on pkg/hive/cell/metric.go:
```diff
@@ -0,0 +1,140 @@
+		))
+	}
+
+	// Lets be strict for now, could lift this in the future if we ever need to
```
```suggestion
	// Let's be strict for now, could lift this in the future if we ever need to
```

qmonnet commented on pkg/metrics/metric/counter.go:
```diff
@@ -0,0 +1,167 @@
+}
+
+// Add adds the given value to the counter. It panics if the value is <
+// 0.
```
```suggestion
// Add adds the given value to the counter. It panics if the value is < 0.
```

qmonnet commented on pkg/hive/cell/metric.go:
```diff
@@ -0,0 +1,140 @@
+		if !field.Type.Implements(withMetaTyp) {
+			panic(fmt.Errorf(
+				"The struct returned by the constructor passed to cell.Metric has a field '%s', which is not "+
+					"metric.WithMetadata.",
```
Let's not break the string? Same comment for the block below.

qmonnet reviewed (approved):
Doc changes look good, thanks!

The few nits below are not blocking.

nathanjsweet reviewed (approved):
LGTM for cli
